### PR TITLE
Add RLM parser

### PR DESCRIPTION
**Pull request by julianaklulo**
#### What
Add RLM parser and unit test.

#### Why
We need to add suport for RLM license server.

`Task`: https://app.clickup.com/t/18022949/LM-128

### CODE REVIEW

File: agent/lm_agent/parsing/rlm.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+TOTALS_LINE = rf"^(?P<license_feature>\w+) {VERSION}: (?P<user_name>\w+)@(?P<lead_host>{HOSTNAME}) (?P<license_used_by_host>\d+).*$"  # noqa
+DATA_LINE = rf"^\s*(?P<license_feature>\w+) {VERSION}$"
+COUNT_LINE = r"^\s*count: (?P<count>\d+).*inuse: (?P<in_use>\d+).*$"
+
+RX_TOTAL = re.compile(rf"{TOTALS_LINE}")
+RX_DATA = re.compile(rf"{DATA_LINE}")
+RX_COUNT = re.compile(rf"{COUNT_LINE}")
+
+
+def _get_start_offset(lines) -> int:
```
Comment (dusktreader) [inline agent/lm_agent/parsing/rlm.py:21]: I think it would be good to have a few unit tests for this function specifically. The arithmetic is very specific, and we should hammer on a few edge-cases.

File: agent/lm_agent/parsing/rlm.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+VERSION = rf"v{INT}\.{INT}"
+HOSTWORD = r"[a-zA-Z0-9-]+"
+HOSTWORD2 = r"[a-zA-Z0-9-.]+"
+HOSTNAME = rf"{HOSTWORD}(\.{HOSTWORD2})*"
+
+TOTALS_LINE = rf"^(?P<license_feature>\w+) {VERSION}: (?P<user_name>\w+)@(?P<lead_host>{HOSTNAME}) (?P<license_used_by_host>\d+).*$"  # noqa
+DATA_LINE = rf"^\s*(?P<license_feature>\w+) {VERSION}$"
+COUNT_LINE = r"^\s*count: (?P<count>\d+).*inuse: (?P<in_use>\d+).*$"
+
+RX_TOTAL = re.compile(rf"{TOTALS_LINE}")
```
Comment (dusktreader) [inline agent/lm_agent/parsing/rlm.py]: These are already raw strings. Why do we need to interpolate them with f-strings again?

File: agent/lm_agent/parsing/rlm.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+RX_DATA = re.compile(rf"{DATA_LINE}")
+RX_COUNT = re.compile(rf"{COUNT_LINE}")
+
+
+def _get_start_offset(lines) -> int:
+    """Get the start offset of the license data."""
```
Comment (dusktreader) [inline agent/lm_agent/parsing/rlm.py:22]: It would be good to show an example of how the string(s) look to relate to the search logic below.

File: agent/lm_agent/parsing/rlm.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    return i
+
+
+def parse(s: str) -> dict:
+    """
+    Parse lines of the license output with regular expressions
+    """
+    parsed_dict: dict = {"uses": [], "total": []}
+    start_offset = _get_start_offset(s.splitlines())
+    lines = [line.strip() for line in s.splitlines()[start_offset:]]
```
Comment (dusktreader) [inline agent/lm_agent/parsing/rlm.py]: Should probably only `splitlines()` once.

File: agent/lm_agent/parsing/rlm.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            data_data = parsed_data.groupdict()
+            parsed_dict["total"].append(
+                {
+                    "feature": data_data["license_feature"],
+                    "total": int(count_data["count"]),
+                    "used": int(count_data["in_use"]),
+                }
+            )
+        else:
+            continue
```
Comment (dusktreader) [inline agent/lm_agent/parsing/rlm.py]: No need for the `else/continue` here.

File: agent/tests/conftest.py [hunk 6]
```diff
@@ -75,3 +75,78 @@ def lm_output():
...
+		obsolete: 0, min_remove: 120, total checkouts: 16
+
+
+	------------------------
+
+	csci license usage status on licser.server.com (port 63133)
+
+	converge_super v3.0: jbemfv@myserver.example.com 29/0 at 11/01 09:01  (handle: 15a) 
+	converge_super v3.0: cdxfdn@myserver.example.com 27/0 at 11/03 10:38  (handle: 128) 
+    converge_super v3.0: jbemfv@myserver.example.com 37/0 at 11/01 09:01  (handle: 15a)
```
Comment (dusktreader) [inline agent/tests/conftest.py]: indent is off.

File: agent/tests/conftest.py [hunk 7]
```diff
@@ -75,3 +75,78 @@ def lm_output():
         "           jbemfv myserver.example.com /dev/tty (v62.2) (myserver.example.com/24200 12507), "
         "start Thu 10/29 8:09, 37 licenses\n"
     )
+
+
+@fixture
+def rlm_output():
+    return """Setting license file path to 10@licserv.server.com
```
Comment (dusktreader) [inline agent/tests/conftest.py:82]: Instead of having to dedent line 83 and 84, you could do this:
```
return textwrap.dedent("""
    Setting license....
    rlmutil v12.2
    Copyright...


        rlm status...
        rml software....
    ...
    """).strip()
```

This looks better to my eyes for python multiline strings that need to be dedented.

**Review by dusktreader — CHANGES_REQUESTED** [review]

File: agent/tests/parsing/test_rlm.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+"""
+Test the rlm parser
+"""
+from pytest import mark
+
+from lm_agent.parsing.rlm import parse
+
+
+@mark.parametrize(
```
Comment (jamesbeedy) [inline agent/tests/parsing/test_rlm.py:9]: 👍

File: agent/tests/conftest.py (hunk 7 above)
Comment (heitorPB) [reply agent/tests/conftest.py:82]: The way it is looks simple to read. I see no reason to use `textwrap.dedent` (or similar) here, as it is quite an obscure feature of the language.

File: agent/lm_agent/parsing/rlm.py (hunk 1 above)
Comment (julianaklulo) [reply agent/lm_agent/parsing/rlm.py:21]: We added the unit test for this function to check a parseable response and a bad one.

File: agent/lm_agent/parsing/rlm.py (hunk 2 above)
Comment (julianaklulo) [reply agent/lm_agent/parsing/rlm.py]: We don't need, fixed it.

File: agent/lm_agent/parsing/rlm.py (hunk 4 above)
Comment (julianaklulo) [reply agent/lm_agent/parsing/rlm.py]: Fixed it.

File: agent/lm_agent/parsing/rlm.py (hunk 5 above)
Comment (julianaklulo) [reply agent/lm_agent/parsing/rlm.py]: Removed it.

File: agent/tests/conftest.py (hunk 6 above)
Comment (julianaklulo) [reply agent/tests/conftest.py]: Fixed it.

File: agent/lm_agent/parsing/rlm.py (hunk 3 above)
Comment (julianaklulo) [reply agent/lm_agent/parsing/rlm.py:22]: We have the string in a fixture: https://github.com/omnivector-solutions/license-manager/blob/rlm_parser/agent/tests/conftest.py#L81-L150
It's too big to put as a comment in the function. The idea is that we only need the last two sections of the output, so we use the hyphens to search for the line we want to use as the offset.

File: agent/tests/conftest.py (hunk 7 above)
Comment (julianaklulo) [reply agent/tests/conftest.py:82]: I agree with @heitorPB.

Comment (dusktreader) [reply agent/tests/conftest.py:82]: It's just a personal preference. I have the same issue with HEREdocs in other languages. I just don't like lines that aren't indented as part of a function.

**Review by dusktreader — APPROVED** [review]
Nice work!
